Provisioning: Implement authorizer for remaining resources

Proximyst opened the pull request:
We'll now limit what resources users have access to.

manually tested admin & viewer.

Proximyst committed:
feat: implement authorizer for remaining resources

Proximyst committed:
fix: don't allow viewers to write files

Proximyst committed:
Merge remote-tracking branch 'origin/grafana-git-ui-sync' into mariell/authorizer

Proximyst committed:
security: harden blob id fetching

ryantxu commented on pkg/registry/apis/provisioning/register.go:
```diff
@@ -175,15 +176,110 @@ func RegisterAPIService(
+						return authorizer.DecisionDeny, "editor role is required for edits", nil
+					}
+
+				case "resources", "sync", "history", "render":
```
is render always allow?  (it is public anyway)

ryantxu commented on pkg/registry/apis/provisioning/register.go:
```diff
@@ -175,15 +176,110 @@ func RegisterAPIService(
+					// Writing files is only allowed fo Editor and higher.
+					isViewer := id.GetOrgRole().Includes(identity.RoleViewer)
+					isEditor := id.GetOrgRole().Includes(identity.RoleEditor)
+					isReadOperation := a.GetVerb() == apiutils.VerbGet || a.GetVerb() == apiutils.VerbList || a.GetVerb() == apiutils.VerbWatch
```
```suggestion
					isReadOperation := a.GetVerb() == apiutils.VerbGet || a.GetVerb()
```
list and watch verbs do not exist for sub-resource

Proximyst replied:
TIL, thanks!

ryantxu commented on pkg/registry/apis/provisioning/register.go:
```diff
@@ -175,15 +176,110 @@ func RegisterAPIService(
+
+			case "stats":
+				// This can leak information one shouldn't necessarily have access to.
+				if id.GetOrgRole().Includes(identity.RoleAdmin) {
+					return authorizer.DecisionAllow, "", nil
+				} else {
+					return authorizer.DecisionDeny, "admin role is required", nil
+				}
```
```suggestion
				if id.GetOrgRole().Includes(identity.RoleAdmin) {
					return authorizer.DecisionAllow, "", nil
				} 
				return authorizer.DecisionDeny, "admin role is required", nil
```

I don't have a strong opinion on this... but roberto usually asks me to change to this style

ryantxu committed:
add integration test for admin vs viewer

Proximyst committed:
feat: only Get is a valid verb for reads in our subresources

Co-Authored-By: Ryan McKinley <ryantxu@gmail.com>

Proximyst committed:
feat: allow render for all requests

Proximyst committed:
refactor: use guards

Not changing code that goes `if a { } else if b { } else { }` as the semantic meaning of the different branches is
easier to parse.

ryantxu reviewed (approved):
LGTM -- lets start with this and figure out how to evolve into real permisisons (not based on role)

Proximyst committed:
Merge remote-tracking branch 'origin/grafana-git-ui-sync' into mariell/authorizer
